controllers/util/conditions.go

```diff
@@ -87,28 +88,108 @@ func ConditionAppend(
+	falseSubConditions := []*metav1.Condition{}
+	unknownSubConditions := []*metav1.Condition{}
 
 	for _, subCondition := range subConditions {
 		if subCondition == nil {
 			continue
 		}
 
-		if subCondition.Status == metav1.ConditionFalse {
-			conditionSet(conditions, *subCondition)
-
-			return
+		switch subCondition.Status {
+		case metav1.ConditionFalse:
+			falseSubConditions = oldestConditions(subCondition, falseSubConditions)
+		case metav1.ConditionUnknown:
+			unknownSubConditions = oldestConditions(subCondition, unknownSubConditions)
+		case metav1.ConditionTrue:
+			trueSubConditions = oldestConditions(subCondition, trueSubConditions)
 		}
+	}
+
+	switch {
+	case len(falseSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(falseSubConditions, ignoreReasons))
+	case len(unknownSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(unknownSubConditions, ignoreReasons))
+	case len(trueSubConditions) != 0:
+		conditionSet(conditions, mergedCondition(trueSubConditions, ignoreReasons))
+	}
+}
+
+// oldestConditions returns a list of conditions that are the same generation and the oldest among newCondition and
+// conditions[]. Initial call should pass an empty conditions[] and iterate over conditions to merge, as the function
+// expects conditions in passed in conditions[] to be the of the same generation.
+func oldestConditions(newCondition *metav1.Condition, conditions []*metav1.Condition) []*metav1.Condition {
+	retConditions := []*metav1.Condition{}
+
+	if len(conditions) == 0 {
+		retConditions = append(retConditions, newCondition)
+
+		return retConditions
+	}
+
+	// All conditions in conditions[] are from an older generation
+	if conditions[0].ObservedGeneration < newCondition.ObservedGeneration {
+		return conditions
+	}
+
+	// newCondition is of the same generation as conditions in conditions[]
+	if conditions[0].ObservedGeneration == newCondition.ObservedGeneration {
+		conditions = append(conditions, newCondition)
 
-		trueSubConditions = append(trueSubConditions, subCondition)
+		return conditions
 	}
 
-	if len(trueSubConditions) > 0 {
-		conditionSet(conditions, *trueSubConditions[len(trueSubConditions)-1])
+	// newCondition generation is lower than all conditions in conditions[]
+	retConditions = append(retConditions, newCondition)
+
+	return retConditions
+}
+
+// mergedCondition merges messages from conditions[] that MUST contain conditions with the same Type,
+// ObservedGeneration, Status.
+// If a condition in conditions[] has a Reason in ignoreReasons it is ignored in the merge unless all conditions
+// have a reason from ignoreReasons
+func mergedCondition(conditions []*metav1.Condition, ignoreReasons []string) metav1.Condition {
+	merged := metav1.Condition{}
+
+	for _, subCondition := range conditions {
+		if merged.Type == "" {
+			merged = *subCondition
+
+			continue
+		}
+
+		if subCondition.Reason == merged.Reason {
+			merged.Message += ". " + subCondition.Message
```

Comment:
Question: Isn't this going to cause the main condition `Message` to grow non-stop if the reason changes every time?

Reply:
The merged condition starts empty and accumulates messages from subConditions in the current iteration. So it does not accumulate messages from older reasons. IOW it is reevaluated each time based on the subConditions passed. Does that address the concern @BenamarMk ?

Reply:
Ah, I see it now, we are not passing the .status.conditions as the first argument, but instead we are passing one of the tmp conditions. All good.